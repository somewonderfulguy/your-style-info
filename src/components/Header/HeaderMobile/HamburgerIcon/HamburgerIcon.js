import React, {memo, useRef} from 'react'
import {bool, func} from 'prop-types'
import {useSpring, animated} from 'react-spring'

import styles from './HamburgerIcon.module.css'

const propTypes = {
  isOpen: bool,
  onClick: func
}

const defaultProps = {
  isOpen: false,
  onClick: () => {}
}

const HamburgerIcon = ({isOpen, onClick}) => {
  const hamburgerDOM = useRef(null)
  const hamburgerHeight = hamburgerDOM.current ? hamburgerDOM.current.clientHeight : 0
  const lineHeight = hamburgerDOM.current ? hamburgerDOM.current.querySelector('div').clientHeight : 0
  const middlePosition = hamburgerHeight / 2 - lineHeight / 2
  const lineColor = getComputedStyle(document.documentElement).getPropertyValue('--mobile-menu-hamburger-color')

  const {timeline} = useSpring({
    config: {duration: 350},
    timeline: isOpen ? 1 : 0
  })

  const range = [0, 0.4, 0.6, 1]

  const topLine = {
    top: timeline.interpolate(range, [0, middlePosition, middlePosition, middlePosition]),
    transform: timeline.interpolate({range, output: [0, 0, 0, 45]}).interpolate(x => `rotate(${x}deg)`)
  }

  const bottomLine = {
    bottom: topLine.top,
    transform: timeline.interpolate({range, output: [0, 0, 0, 45]}).interpolate(x => `rotate(-${x}deg)`)
  }

  const midLine = {
    background: timeline.interpolate({range, output: [lineColor || '#000', 'transparent', 'transparent', 'transparent']})
  }

  return (
    <button ref={hamburgerDOM} onClick={onClick} className={styles.hamburger} type="button">
      <animated.div style={topLine} />
      <animated.div style={midLine} />
      <animated.div style={bottomLine} />
    </button>
  )
}

HamburgerIcon.propTypes = propTypes
HamburgerIcon.defaultProps = defaultProps

export default memo(HamburgerIcon)