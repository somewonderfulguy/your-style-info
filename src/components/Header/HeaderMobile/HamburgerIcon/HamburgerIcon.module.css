.hamburger {
  --mobile-menu-hamburger-size: 24px;
  --mobile-menu-hamburger-line: 2px;
  --mobile-menu-hamburger-duration: 1s;
}

.hamburger {
  width: var(--mobile-menu-hamburger-size);
  height: calc(var(--mobile-menu-hamburger-size) * 0.75);
  position: relative;
  outline: none;

  /* fix positioning */
  display: flex;
}

.hamburgerDark {
  composes: hamburger;
}

.hamburger > div {
  width: 100%;
  height: var(--mobile-menu-hamburger-line);
  position: absolute;
  background-color: #000;
}

.hamburgerDark > div {
  width: 100%;
  height: var(--mobile-menu-hamburger-line);
  position: absolute;
  background-color: lightgray;
}

.hamburger > div:nth-child(1) {
  top: 0;
}
.hamburger > div:nth-child(2) {
  top: calc(50% - var(--mobile-menu-hamburger-line) / 2);
}
.hamburger > div:nth-child(3) {
  bottom: 0;
}
