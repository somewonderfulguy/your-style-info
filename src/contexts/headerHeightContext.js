import React, {createContext, useContext, useMemo, useState} from 'react'

const HeaderHeightContext = createContext()

const useHeaderHeight = () => {
  const context = useContext(HeaderHeightContext)
  if(!context) {
    throw new Error('useHeaderHeight must be used within a HeaderHeightProvider')
  }
  return context
}

const HeaderHeightProvider = props => {
  const [headerHeight, setHeaderHeight] = useState(0)
  const value = useMemo(() => ({headerHeight, setHeaderHeight}), [headerHeight])
  return <HeaderHeightContext.Provider value={value} {...props} />
}

export {HeaderHeightProvider, useHeaderHeight}