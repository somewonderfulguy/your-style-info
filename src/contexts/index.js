export * from './headerHeightContext'
export * from './loadingContext'
export * from './localisationContext'
export * from './screenDimensionsContext'
export * from './themeContext'