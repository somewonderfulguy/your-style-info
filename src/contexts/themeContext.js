import React, {createContext, useCallback, useContext, useMemo, useEffect, useState} from 'react'

const ThemeContext = createContext()

const useTheme = () => {
  const context = useContext(ThemeContext)
  if(!context) {
    throw new Error('useTheme must be used within a ThemeProvider')
  }
  const [isDarkTheme, setDarkTheme] = context

  const switchTheme = useCallback(
    () => setDarkTheme(!isDarkTheme),
    [isDarkTheme, setDarkTheme]
  )

  return {isDarkTheme, switchTheme}
}

const ThemeProvider = props => {
  const preferDarkQuery = '(prefers-color-scheme: dark)'
  const [isDarkTheme, setDarkTheme] = useState(
    () =>
      window.localStorage.getItem('isDarkTheme') === 'true' ||
      (window.matchMedia(preferDarkQuery).matches ? true : false),
  )

  useEffect(() => {
    const mediaQuery = window.matchMedia(preferDarkQuery)
    const handleChange = () => setDarkTheme(mediaQuery.matches ? true : false)
    mediaQuery.addListener(handleChange)
    return () => mediaQuery.removeListener(handleChange)
  }, [])

  useEffect(() => {
    window.localStorage.setItem('isDarkTheme', isDarkTheme)
  }, [isDarkTheme])

  const value = useMemo(() => [isDarkTheme, setDarkTheme], [isDarkTheme])

  return <ThemeContext.Provider value={value} {...props} />
}

export {ThemeProvider, useTheme}