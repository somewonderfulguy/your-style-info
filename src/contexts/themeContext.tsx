import React, {
  createContext, Dispatch, ReactNode, SetStateAction,
  useCallback, useContext, useEffect, useState
} from 'react'

type dispatchType = Dispatch<SetStateAction<boolean>>

const ThemeStateContext = createContext<boolean | undefined>(undefined)
ThemeStateContext.displayName = 'ThemeStateContext'

const ThemeDispatchContext = createContext<dispatchType | undefined>(undefined)
ThemeDispatchContext.displayName = 'ThemeDispatchContext'

const ThemeProvider = ({ children }: { children: ReactNode | ReactNode[] }) => {
  const preferDarkQuery = '(prefers-color-scheme: dark)'
  const [isDarkTheme, setDarkTheme] = useState(
    () =>
      window.localStorage.getItem('isDarkTheme') === 'true' ||
      (window.matchMedia(preferDarkQuery).matches ? true : false)
  )

  useEffect(() => {
    const mediaQuery = window.matchMedia(preferDarkQuery)
    const handleChange = () => setDarkTheme(mediaQuery.matches ? true : false)
    mediaQuery.addListener(handleChange)
    return () => mediaQuery.removeListener(handleChange)
  }, [])

  useEffect(() => {
    window.localStorage.setItem('isDarkTheme', String(isDarkTheme))
  }, [isDarkTheme])

  return (
    <ThemeStateContext.Provider value={isDarkTheme}>
      <ThemeDispatchContext.Provider value={setDarkTheme}>
        {children}
      </ThemeDispatchContext.Provider>
    </ThemeStateContext.Provider>
  )
}

const useThemeState = () => {
  const isDarkTheme = useContext(ThemeStateContext)

  if(isDarkTheme === undefined) {
    throw new Error('useThemeState must be used within a ThemeProvider')
  }

  return isDarkTheme
}

const useThemeSwitch = () => {
  const setDarkTheme = useContext(ThemeDispatchContext)

  if(setDarkTheme === undefined) {
    throw new Error('useThemeSwitch must be used within a ThemeProvider')
  }

  const switchTheme = useCallback(
    () => setDarkTheme(prevTheme => !prevTheme),
    [setDarkTheme]
  )

  return switchTheme
}

export {ThemeProvider, useThemeState, useThemeSwitch}