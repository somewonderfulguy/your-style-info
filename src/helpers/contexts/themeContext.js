import React, {createContext, useContext, useMemo, useEffect, useState} from 'react'

const ThemeContext = createContext()

const useTheme = () => {
  const context = useContext(ThemeContext)
  if(!context) {
    throw new Error('useTheme must be used within a ThemeProvider')
  }
  const [isDarkTheme, setDarkTheme] = context

  const switchTheme = () => {
    setDarkTheme(!isDarkTheme)
    localStorage.setItem('isDarkTheme', !isDarkTheme)
  }

  return {isDarkTheme, switchTheme}
}

const ThemeProvider = props => {
  const [isDarkTheme, setDarkTheme] = useState(false)
  const [hasThemeLoaded, setThemeLoaded] = useState(false)

  useEffect(() => {
    const isLocalStorageDark = localStorage.getItem('isDarkTheme')
    setDarkTheme(isLocalStorageDark === 'true')
    setThemeLoaded(true)
  }, [])

  const value = useMemo(() => [isDarkTheme, setDarkTheme], [isDarkTheme])

  if(!hasThemeLoaded) {
    // avoiding render default light theme and then re-render to
    // dark theme if localStorage keeps dark mode as true
    return <div />
  }

  return <ThemeContext.Provider value={value} {...props} />
}

export {ThemeProvider, useTheme}