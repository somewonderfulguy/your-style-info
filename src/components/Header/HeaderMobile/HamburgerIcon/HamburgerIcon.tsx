import { memo, useRef } from 'react'
import { useSpring, animated } from 'react-spring'

import { useThemeState } from '~contexts/themeContext'

import styles from './HamburgerIcon.module.css'

type Props = {
  isOpen?: boolean
  onClick?: () => void
}

const HamburgerIcon = ({ isOpen = false, onClick = () => {} }: Props) => {
  const isDarkTheme = useThemeState()
  const hamburgerDOM = useRef<HTMLButtonElement>(null)
  const hamburgerHeight = hamburgerDOM.current?.clientHeight ?? 0
  const lineHeight =
    hamburgerDOM.current?.querySelector('div')?.clientHeight ?? 0
  const middlePosition = hamburgerHeight / 2 - lineHeight / 2
  const lineColor = isDarkTheme ? 'lightgray' : '#000'

  const { timeline } = useSpring({
    config: { duration: 350 },
    timeline: isOpen ? 1 : 0
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  }) as any

  const range = [0, 0.4, 0.6, 1]

  const topLine = {
    top: timeline.interpolate(range, [
      0,
      middlePosition,
      middlePosition,
      middlePosition
    ]),
    transform: timeline
      .interpolate({ range, output: [0, 0, 0, 45] })
      .interpolate((x) => `rotate(${x}deg)`)
  }

  const bottomLine = {
    bottom: topLine.top,
    transform: timeline
      .interpolate({ range, output: [0, 0, 0, 45] })
      .interpolate((x) => `rotate(-${x}deg)`)
  }

  const midLine = {
    background: timeline.interpolate({
      range,
      output: [lineColor, 'transparent', 'transparent', 'transparent']
    })
  }

  return (
    <button
      ref={hamburgerDOM}
      onClick={onClick}
      className={styles[isDarkTheme ? 'hamburgerDark' : 'hamburger']}
      type="button"
      title="Navigation"
    >
      <animated.div style={topLine} />
      <animated.div style={midLine} />
      <animated.div style={bottomLine} />
    </button>
  )
}

export default memo(HamburgerIcon)
