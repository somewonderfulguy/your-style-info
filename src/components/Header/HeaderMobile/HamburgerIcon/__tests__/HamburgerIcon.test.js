import React from 'react'
import {render} from '@testing-library/react'

import * as mockThemeContext from 'contexts/themeContext'
import HamburgerIcon from '..'

jest.spyOn(mockThemeContext, 'useTheme').mockImplementation(() => ({isDarkTheme: false}))
afterEach(() => jest.clearAllMocks())

test('snapshot diffrenece between default and clicked states', async () => {
  const {asFragment, rerender} = render(<HamburgerIcon />)
  const defaultState = asFragment()
  rerender(<HamburgerIcon isOpen />)

  // TODO remove once react-spring 9.0.0 released
  await new Promise((r) => setTimeout(r, 400))

  const clickedState = asFragment()

  expect(defaultState).toMatchDiffSnapshot(clickedState, {
    contextLines: 10,
    aAnnotation: 'default state (three lines)',
    bAnnotation: 'clicked state (cross)'
  })
  expect(mockThemeContext.useTheme).toHaveBeenCalledTimes(2)

  // TODO close and compare
})

test('snapshot difference in dark/light theme', async () => {
  const lightTheme = render(<HamburgerIcon />).asFragment()

  mockThemeContext.useTheme.mockReturnValueOnce({isDarkTheme: true})
  const darkTheme = render(<HamburgerIcon />).asFragment()

  expect(lightTheme).toMatchDiffSnapshot(darkTheme, {
    contextLines: 1,
    aAnnotation: 'light theme',
    bAnnotation: 'dark theme'
  })

  expect(mockThemeContext.useTheme).toHaveBeenCalledTimes(2)

  // TODO switch back and compare
})