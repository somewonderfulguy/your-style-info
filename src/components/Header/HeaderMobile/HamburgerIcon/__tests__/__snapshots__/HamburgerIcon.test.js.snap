// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot difference in dark/light theme 1`] = `
Snapshot Diff:
- light theme
+ dark theme

@@ -2,3 +2,3 @@
    <button
-     class="hamburger"
+     class="hamburgerDark"
      type="button"
@@ -9,3 +9,3 @@
      <div
-       style="background: rgb(0, 0, 0);"
+       style="background: rgb(211, 211, 211);"
      />
`;

exports[`snapshot diffrenece between default and clicked states 1`] = `
Snapshot Diff:
- default state (three lines)
+ clicked state (cross)

  <DocumentFragment>
    <button
      class="hamburger"
      type="button"
    >
      <div
-       style="top: 0px; transform: rotate(0deg);"
+       style="top: 0px; transform: rotate(45deg);"
      />
      <div
-       style="background: rgb(0, 0, 0);"
+       style="background: rgba(0, 0, 0, 0);"
      />
      <div
-       style="bottom: 0px; transform: rotate(-0deg);"
+       style="bottom: 0px; transform: rotate(-45deg);"
      />
    </button>
  </DocumentFragment>
`;
